REFRESH's automated cluster scheduling -- planning, sequencing, catalog object change

ggevay opened the pull request:
This is the 2. PR of the automated cluster scheduling for `REFRESH`, implementing the planning, sequencing, and the catalog object change. (1. was the parsing: https://github.com/MaterializeInc/materialize/pull/25998)

(The main part of the implementation, i.e., actually turning clusters on and off, will come in the next PR.)

### Motivation

  * This PR is a step towards adding a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/25712

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3586,12 +3586,18 @@ impl<T: AstInfo> AstDisplay for RefreshOptionValue<T> {
+    fn default() -> Self {
+        ClusterScheduleOptionValue::Manual
+    }
+}
```
As a general rule we don't impl `Default` for AST types, because they are not meant to be constructed directly, and the "defaults" flow from parser.rs and the SQL planner. Does it cause any friction to leave `Default` unimplemented here, and declare the default in the `generate_extracted_config!` macro as was done previously?

benesch replied:
Oh, I see, `ALTER RESET`. Ok, never mind.

ggevay commented:
[CI failed](https://buildkite.com/materialize/tests/builds/79044) on some catalog migration things that I didn't know about. I fixed the things that tests were complaining about, and tests are passing now, but could you please check the last commit? In particular, in `v49_to_v50.rs` is my thinking correct that I don't need to write any code for the migration because I'm just adding a new field whose default value will be fine for old catalog objects?

ParkMyCar commented:
> [CI failed](https://buildkite.com/materialize/tests/builds/79044) on some catalog migration things that I didn't know about. I fixed the things that tests were complaining about, and tests are passing now, but could you please check the last commit? In particular, in `v49_to_v50.rs` is my thinking correct that I don't need to write any code for the migration because I'm just adding a new field whose default value will be fine for old catalog objects?

Tests for the win! Unfortunately you do need a migration here, specifically because there is a default value. Otherwise the values won't consolidate correctly since when serialized the new version will include the default value. I would look at the [v45_to_v46 migration](https://github.com/MaterializeInc/materialize/blob/main/src/catalog/src/durable/upgrade/v45_to_v46.rs) as an example. The `wire_compatible!(...)` macro should cut down on the boilerplate. Specifically `ReplicaMergeEffort` and `ReplicaLogging` are wire compatible across v49 and v50. More than happy to hop in a Huddle tomorrow if you run into any issues!

ggevay commented:
Is there some documentation somewhere that gives a high-level overview of catalog migrations and explains what a catalog migration needs to do? I've added some more code, but I have several questions:

A side question: When I added the REFRESH options on materialized views [some time ago](https://github.com/MaterializeInc/materialize/pull/23870/files#diff-8d2fe0db05338dd8df0efaffc2dbdeb4075ccf16e70f98ffa5911c75b0900974), that catalog migration was different: I used a visitor on the AST of the CREATE MATERIALIZED VIEW statement instead of manipulating protobuf things. Is this because the durable representation of materialized views is simply the SQL string?

In my current situation, where I'm changing `ClusterVariantManaged` by adding a new field that has a default value, at first I naively thought that we can simply deserialize the old on-disk representation into the new in-memory representation (automatically filling in the default value), and then at some point we serialize the new in-memory representation into the new on-disk representation, and just overwrite the catalog on the disk, and we are done. Are things more incremental somehow, so if we don't have an explicit migration then we don't notice that the on-disk representation should be overwritten with the new representation, and the old on-disk representation is just left there? Also, you mention above that "values won't consolidate correctly" if I don't add a migration. When do these things have to consolidate? When dropping the cluster?

What does the `wire_compatible` macro do and why? I'm probably missing some nuances here, but I thought that compatibility just means that if I serialize an old struct and then try to deserialize it into a new struct, then the deserialization succeeds and yields a semantically correct struct (in my case, by filling in the default value for the added field). But, from reading the code, it seems the `wire_compatible` macro uses the term to mean something stronger:
1. The proptest added by the macro converts from old to new, and serializes both the old and new structs and asserts that the serialized representations are byte-wise equal.
2. Naively, I thought that we just need to convert in one direction, from the old representation to the new, but it seems that the macro adds the `convert` method in both ways, and also checks 1. when converting in both directions. Why is this? (If I remember correctly, we don't support downgrades, and not planning to do so.)

[I ran Nightly](https://buildkite.com/materialize/nightlies/builds/7061#018e7aeb-c838-48dd-9f41-af917c082809), and there are still some assert fails in upgrade tests. I'll keep digging, but I guess someone who knows how catalog migrations work  might be able to figure this out much more quickly.

ggevay commented:
Had a zoom with @ParkMyCar, and we discussed the above questions, and then look into the assert fail, and he quickly pointed out that my `Default::default()` was invoking `Option`'s default, and not my struct's default... Gonna kick off another nightly now.

ggevay commented:
The [Nightly](https://buildkite.com/materialize/nightlies/builds/7076) upgrade tests have succeeded this time! (The "Output consistency" failures [are unrelated](https://materializeinc.slack.com/archives/C01LKF361MZ/p1711455863653069?thread_ts=1711454798.546479&cid=C01LKF361MZ) to this PR.)

Squashed the last few commits, and setting to auto-merge.
